lib/src/revset.rs

```diff
@@ -691,6 +697,13 @@ static BUILTIN_FUNCTION_MAP: Lazy<HashMap<&'static str, RevsetFunction>> = Lazy:
             pattern,
         )))
     });
+    map.insert("author_date", |function, context| {
+        let [arg] = function.expect_exact_arguments()?;
+        let pattern = expect_timeexpression(arg, context.now().to_owned())?;
+        Ok(RevsetExpression::filter(RevsetFilterPredicate::AuthorDate(
+            pattern,
+        )))
+    });
```

Comment:
It's unclear that `author_date(x)` means `t >= x`. Maybe it can be renamed to `author_date_after()` or `authored_after()`?

I also like the git-branchless' syntax. `author_date(after:x)`.
https://github.com/martinvonz/jj/issues/3287

Reply:
I'm open to renaming it to `X_after()`.

I originally thought the syntax was going to end up like one of these:

- `author_date(after:"2024-03-01")`
- `author_date(after="2024-03-01")`
- `author_date("after 2024-03-01")`

However, I thought this introduced a problem: It implies that `before:` should work too. And "before" has design decisions that I thought were tough to resolve in a way that satisfies most people.

- Does `author_date(before="2024")` mean `t < 2024-01-01 00:00:00` or `t < 2025-01-01 00:00:00`? The former is consistent, but a lot of people seem to expect dates to be inclusive on both ends. For example, I think a lot of people expect this to include all of March, whereas I expect it to exclude March 31: `date(after:"2024-03-01") & date(before:"2024-03-31")`

- If "before" is inclusive, then partial dates should resolve to the _end_ of their matching times:
  | revset | after | before |
  | -- | -- | -- |
  | `after:"2024"` | `after:"2024-01-01 00:00:00"` | `before:"2024-12-31 23:59:59"` |
  | `after:"2024-03-01"` | `after:"2024-03-01 00:00:00"` | `before:"2024-03-01 23:59:59"` |
  | `after:"March 1"` | `after:"2024-03-01 00:00:00"` | `before:"2024-03-01 23:59:59"` |

  chrono-english doesn't support this as far as I can tell, partial dates always resolve to the beginning of the time period. And there's no way after parsing to learn what precision was given, so we'd have to write our own parser. I originally did this with Pest, but supporting localized date formats like "April 1" vs. "1 April" seemed like it was going to be a losing battle.

Because I wasn't sure what to do, I opted to omit the word "after" entirely and just make it the default, to avoid implying that "before" should be there.

So, what do you think? Add the word "after" and then deal with feature requests for "before"?

Reply:
> * Does `author_date(before="2024")` mean `t < 2024-01-01 00:00:00` or `t < 2025-01-01 00:00:00`? The former is consistent, but a lot of people seem to expect dates to be inclusive on both ends. For example, I think a lot of people expect this to include all of March, whereas I expect it to exclude March 31: `date(after:"2024-03-01") & date(before:"2024-03-31")`

Appears that git-branchless do the former (@arxanas correct me if I'm wrong), but I agree I belong to the party who expects the latter "inclusive" date.

> chrono-english doesn't support this as far as I can tell, partial dates always resolve to the beginning of the time period.

Correct. There's no public API to extract inner parse result afaict.

Mercurial does brute-force parsing by trying srtftime formats in order. Maybe we can do the same without writing full-blown parser?

https://repo.mercurial-scm.org/hg/file/tip/mercurial/utils/dateutil.py#l42
https://docs.rs/chrono/latest/chrono/struct.DateTime.html#method.parse_from_str

I think it's not uncommon to look for old commits, so I expect we'll eventually want "before" query. It can be expressed as negated expression, but it has cognitive load.

Reply:
I like the `before:`/`after:` syntax, in line with string patterns.

I think the exclusive interpretation of `before:"2024"` makes more sense: in English, it doesn’t make sense to say an event that happened in the middle of 2024 happened “before 2024”. My intuition expects `after="2024"` to mean 2025 onwards, too, but it’s possible that interpreting it as “after and including the beginning of 2024“ (i.e., “not before 2024”) is both more useful and easier to implement. I think that interpreting “after” that way stretches language less than an inclusive interpretation of “before” (you can imagine that `"2024"` stands for an infinitesimal instant right at the border of the start of 2024, after `2023-12-31T23:59:59.999Z` but before `2024-01-01T00:00:00.000Z`).

Reply:
I implemented `before:` and `after:`. I left `author_date("D")` as a shorthand for `author_date(after:"D")`.